### PR TITLE
[Model] support modernbert 

### LINKED ISSUE #11347: [New Model]: answerdotai/ModernBERT-large
### The model to consider.

https://huggingface.co/answerdotai/ModernBERT-large

### The closest model vllm already supports.

BAAI/bge-base-en-v1.5

### What's your difficulty of supporting the model you want?

_No response_

### Before submitting a new issue...

- [ ] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

### PR DESCRIPTION
**Pull request by xsank**
Support modernBert, test passed on Alibaba-NLP/gte-reranker-modernbert-base.

FIX https://github.com/vllm-project/vllm/issues/11347

eg:
```shell
curl -X 'POST' \
  'http://127.0.0.1:8000/v1/score' \
  -H 'Content-Type: application/json' \
  -d '{
  "model": "Alibaba-NLP/gte-reranker-modernbert-base",
  "text_1": ["what is the capital of China?","how to implement quick sort in python?","how to implement quick sort in python?"],
  "text_2": ["Beijing","Introduction of quick sort","The weather is nice today"]
}'
{"id":"score-808110e90ba2472f904721ccd20034ce","object":"list","created":1744705317,"model":"Alibaba-NLP/gte-reranker-modernbert-base","data":[{"index":0,"object":"score","score":0.89453125},{"index":1,"object":"score","score":0.92138671875},{"index":2,"object":"score","score":0.1572265625}],"usage":{"prompt_tokens":43,"total_tokens":43,"completion_tokens":0,"prompt_tokens_details":null}}
```

Signed-off-by: xsank [xsank@foxmail.com](mailto:xsank@foxmail.com)

### CODE REVIEW

**Comment by xsank** [comment]
@DarkLight1337 , I do not know why the `lint-and-deploy` stage failed, nor can i find any useful information from the detail log...

File: vllm/model_executor/models/modernbert.py [hunk 1]
```diff
@@ -0,0 +1,406 @@
...
+
+        rope_theta = config.global_rope_theta
+        if self.local_attention != (
+                -1, -1) and config.local_rope_theta is not None:
+            rope_theta = config.local_rope_theta
+        self.rotary_emb = ModernBertRotaryEmbedding(config=config,
+                                                    head_size=self.head_dim,
+                                                    dim=self.head_dim,
+                                                    base=rope_theta)
+        self.Wo = nn.Linear(config.hidden_size,
```
Comment (DarkLight1337) [inline vllm/model_executor/models/modernbert.py:147]: Let's try to use parallelized linear layers  (e.g. `RowParallelLinear`, `ColumnParallelLinear`) if possible

File: vllm/model_executor/models/modernbert.py [hunk 2]
```diff
@@ -0,0 +1,406 @@
...
+class ModernBertMLP(nn.Module):
+
+    def __init__(self, config: ModernBertConfig):
+        super().__init__()
+        self.config = config
+        self.Wi = nn.Linear(config.hidden_size,
+                            int(config.intermediate_size) * 2,
+                            bias=config.mlp_bias)
+        self.act = GELUActivation()
+        self.drop = nn.Dropout(config.mlp_dropout)
```
[the remark this answers is not in the document]
Comment (xsank) [reply vllm/model_executor/models/modernbert.py:207]: confg.mlp_dropout is 0.0, which means the dropout not work actually, remove the code is fine.

File: vllm/model_executor/models/modernbert.py (hunk 1 above)
Comment (xsank) [reply vllm/model_executor/models/modernbert.py:147]: ok, let me see how to do with this case.

File: vllm/model_executor/models/modernbert.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (xsank) [reply vllm/model_executor/models/modernbert.py]: commit 0c16ece solved this.

File: tests/models/registry.py [hunk 3]
```diff
@@ -273,6 +273,9 @@ def check_available_online(
     "BertForSequenceClassification": _HfExamplesInfo("cross-encoder/ms-marco-MiniLM-L-6-v2"),  # noqa: E501
     "RobertaForSequenceClassification": _HfExamplesInfo("cross-encoder/quora-roberta-base"),  # noqa: E501
     "XLMRobertaForSequenceClassification": _HfExamplesInfo("BAAI/bge-reranker-v2-m3"),  # noqa: E501
+    "ModernBertForSequenceClassification":
+        _HfExamplesInfo("Alibaba-NLP/gte-reranker-modernbert-base",
+                        min_transformers_version="4.49"),
```
Comment (DarkLight1337) [inline tests/models/registry.py:278]: Also can you add this model to the Supported Models page in the docs?

Comment (xsank) [reply tests/models/registry.py:278]: Done

File: vllm/model_executor/models/modernbert.py (hunk 1 above)
Comment (xsank) [reply vllm/model_executor/models/modernbert.py]: @DarkLight1337 the linear layers was replaced, but the attention layer not works so well on the modernBert case. for the score is wrong.

Comment (DarkLight1337) [reply vllm/model_executor/models/modernbert.py]: Make sure you're using the correct attention type.

Comment (xsank) [reply vllm/model_executor/models/modernbert.py]: Thanks you, i should use the encoder only type. f7dd19b solved this. And i should dive into the code and learn more.

File: vllm/model_executor/models/modernbert.py [hunk 4]
```diff
@@ -0,0 +1,351 @@
...
+        super().__init__()
+        config = vllm_config.model_config.hf_config
+        self.config = config
+        self.embeddings = ModernBertEmbeddings(config)
+        self.encoder_layer = ModernBertEncoderLayer(vllm_config)
+        self.final_norm = nn.LayerNorm(config.hidden_size,
+                                       eps=config.norm_eps,
+                                       bias=config.norm_bias)
+        self.gradient_checkpointing = False
+        self.dtype = torch.float16
```
Comment (DarkLight1337) [inline vllm/model_executor/models/modernbert.py:239]: These lines are unnecessary

File: vllm/model_executor/models/modernbert.py [hunk 5]
```diff
@@ -0,0 +1,351 @@
...
+
+class ModernBertMLP(nn.Module):
+
+    def __init__(self, config: ModernBertConfig):
+        super().__init__()
+        self.config = config
+        self.Wi = nn.Linear(config.hidden_size,
+                            int(config.intermediate_size) * 2,
+                            bias=config.mlp_bias)
+        self.act = GELUActivation()
```
[the remark this answers is not in the document]
Comment (xsank) [reply vllm/model_executor/models/modernbert.py]: Done

File: vllm/model_executor/models/modernbert.py (hunk 4 above)
Comment (xsank) [reply vllm/model_executor/models/modernbert.py]: Done

File: vllm/model_executor/models/modernbert.py [hunk 6]
```diff
@@ -0,0 +1,351 @@
...
+                                          bias=config.norm_bias)
+        self.attn = ModernBertAttention(config=config, layer_id=layer_id)
+        self.mlp_norm = nn.LayerNorm(config.hidden_size,
+                                     eps=config.norm_eps,
+                                     bias=config.norm_bias)
+        self.mlp = ModernBertMLP(config)
+
+    @torch.compile(dynamic=True)
+    def compiled_mlp(self, hidden_states: torch.Tensor) -> torch.Tensor:
+        return self.mlp(self.mlp_norm(hidden_states))
```
[the remark this answers is not in the document]
Comment (DarkLight1337) [reply vllm/model_executor/models/modernbert.py]: Still you should add `support_torch_compile` to the model (see other models for examples)

**Review by DarkLight1337 — APPROVED** [review]
Thanks for adding this model!

**Comment by xsank** [comment]
> hi @xsank thanks for adding this. what are the options to run the modern bert model? if i look in https://huggingface.co/Alibaba-NLP/gte-reranker-modernbert-base/tree/main - there is model.safetensors (tensorflow) and onnx.
> 
> pr will also work with https://huggingface.co/answerdotai/ModernBERT-base ?

In fact, this commit support the common architectures of `ModernBertForSequenceClassification`, you could have a try. if you found any thing not fit well, please tell me.

**Comment by geraldstanje** [comment]
@xsank i fine tuned [answerdotai/ModernBERT-base](https://huggingface.co/answerdotai/ModernBERT-base) as a binary text classifier. does that fall under ModernBertForSequenceClassification? the pr you created will return the logits and i need to do the softmax myself?

**Comment by xsank** [comment]
> @xsank @lionelvillard i fine tuned [answerdotai/ModernBERT-base](https://huggingface.co/answerdotai/ModernBERT-base) as a binary text classifier - does that fall under ModernBertForSequenceClassification? edit: i see `"architectures": ["ModernBertForMaskedLM"]` in https://huggingface.co/answerdotai/ModernBERT-base/blob/main/config.json#L4 - so i assume we need another pr to support ModernBertForMaskedLM?
> 
> for when you support architecture "ModernBertForMaskedLM" will you need will you need to call func softmax for the logits returned?

It seems that moderBert series model have some tiny difference, let me see how to support all the features in the same class. Finally, i feel so sorry that i made a mistake, which i only support a part, there is still some work todo.

**Comment by geraldstanje** [comment]
@xsank thanks for your reply - could you also add it for architectore ModernBertForMaskedLM? that would be amazing.

**Comment by geraldstanje** [comment]
@xsank any updates?

**Comment by geraldstanje** [comment]
hi @xsank would you be able to add it for ModernBertForMaskedLM too?

**Comment by geraldstanje** [comment]
hi @xsank can you also add ModernBertForMaskedLM?

**Comment by geraldstanje** [comment]
hi @mhillebrand the pooling is needed to support /classify and has not been added to V0 and V1?

**Comment by DarkLight1337** [comment]
Yes, `/classify` is for classification models which is a subset of pooling models.

**Comment by DarkLight1337** [comment]
You might need to set `--task classify` to use the model as a classification model though.

**Comment by geraldstanje** [comment]
@DarkLight1337 yes + for online serving use openai vllm container from nightly build.

**Comment by DarkLight1337** [comment]
Can you show the command you used to serve the model?

**Comment by geraldstanje** [comment]
@DarkLight1337 using
```
curl -s -X POST http://localhost:8000/score \
  -H "Content-Type: application/json" \
  -d '{"text_1": ["I love machine learning"], "text_2": ["foo"]}'
```

**Comment by DarkLight1337** [comment]
@xsank can you help look into this?

**Comment by geraldstanje** [comment]
@DarkLight1337 is there a way to add a post processing step using a threshold before deciding the class?

**Comment by DarkLight1337** [comment]
No, to achieve that I suggest getting the logits/probabilities and processing them by yourself

**Comment by geraldstanje** [comment]
you refer to using the probs from the classify output?
```
{
  "id": "classify-d72c9764b4cd476081c7d336a26ed724",
  "object": "list",
  "created": 1748063869,
  "model": "/model",
  "data": [
    {
      "index": 0,
      "label": "LABEL_0",
      "probs": [
        5.93359375,
        -6.9453125
      ],
      "num_classes": 2
    }
  ],
  "usage": {
    "prompt_tokens": 6,
    "total_tokens": 6,
    "completion_tokens": 0,
    "prompt_tokens_details": null
  }
}
```

**Comment by geraldstanje** [comment]
@DarkLight1337 alternative would be to add a classifyAndApplyThreshold to vllm and rebuild the container...

**Comment by noooop** [comment]
@xsank 

The results of the Alibaba-NLP/gte-reranker-modernbert-base model seem to slight differ from sentence-transformers. 


```python
import pytest
import torch
from sentence_transformers import CrossEncoder

from vllm import LLM

model_name = "Alibaba-NLP/gte-reranker-modernbert-base"

st_model = CrossEncoder(model_name,
                        model_kwargs={"torch_dtype": torch.float32})
vllm_model = LLM(model_name, task="score", dtype="float32")

sentences = [
    ("ping", "pong"),
    ("ping", "pong" * 16),
    ("ping", "pong" * 24),
    ("ping", "pong" * 32),
    ("ping", "pong" * 48),
    ("ping", "pong" * 64),
    ("ping", "pong" * 128),
]

st_scores = st_model.predict(sentences)

texts_1 = [x[0] for x in sentences]
texts_2 = [x[1] for x in sentences]
outputs = vllm_model.score(texts_1, texts_2)
vllm_scores = [output.outputs.score for output in outputs]


def test_close(s1, s2):
    return float(s1) == pytest.approx(float(s2), rel=0.01)

print(
    [test_close(st_scores[i], vllm_scores[i]) for i in range(len(st_scores))])
```

output

```
[True, True, True, True, False, False, False]
```
